pkg/network/ebpf/c/protocols/tls/tls.h

```diff
@@ -11,122 +10,420 @@
     }
 
-    // Copying struct to the stack, to avoid modifying the original buffer that will be used for other classifiers.
-    tls_record_header_t tls_record_header = *(tls_record_header_t *)buf;
-    // Converting the fields to host byte order.
-    tls_record_header.version = bpf_ntohs(tls_record_header.version);
-    tls_record_header.length = bpf_ntohs(tls_record_header.length);
+    return 0;
+}
+
 
-    // Checking the version in the record header.
-    if (!is_valid_tls_version(tls_record_header.version)) {
-        return false;
+static __always_inline int parse_tls_extensions(struct __sk_buff *skb, __u64 *offset, __u64 extensions_end, __u32 skb_len, tls_info_t *tags, bool is_client_hello) {
+    __u16 extension_type;
+    __u16 extension_length;
+
+    #pragma unroll(MAX_EXTENSIONS)
+    for (int i = 0; i < MAX_EXTENSIONS; i++) {
+        if (*offset + 4 > extensions_end) {
+            break;
+        }
+        // Read Extension Type (2 bytes)
+        if (bpf_skb_load_bytes(skb, *offset, &extension_type, sizeof(extension_type)) < 0)
+            return -1;
+        extension_type = bpf_ntohs(extension_type);
+        *offset += 2;
+
+        // Read Extension Length (2 bytes)
+        if (bpf_skb_load_bytes(skb, *offset, &extension_length, sizeof(extension_length)) < 0)
+            return -1;
+        extension_length = bpf_ntohs(extension_length);
+        *offset += 2;
+
+        // Ensure we don't read beyond the packet
+        if (*offset + extension_length > skb_len || *offset + extension_length > extensions_end)
+            return -1;
+
+        // Check for supported_versions extension
+        if (extension_type == SUPPORTED_VERSIONS_EXTENSION) {
+            int res = parse_supported_versions_extension(skb, offset, extensions_end, skb_len, tags, is_client_hello);
+            if (res != 0)
+                return res;
+        } else {
+            // Skip other extensions
+            *offset += extension_length;
+        }
+
+        // Ensure we don't run past the extensions_end
+        if (*offset >= extensions_end)
+            break;
     }
 
-    // Checking the length in the record header is not greater than the maximum payload length.
-    if (tls_record_header.length > TLS_MAX_PAYLOAD_LENGTH) {
-        return false;
+    return 0;
+}
+
+
+// parse_client_hello reads the ClientHello message from the TLS handshake and populates select tags
+static __always_inline int parse_client_hello(struct __sk_buff *skb, __u64 offset, __u32 skb_len, tls_info_t *tags) {
+    __u32 handshake_length;
+    __u16 client_version;
+    int res;
+
+    // Parse the handshake header
+    res = parse_tls_handshake_header(skb, &offset, skb_len, &handshake_length, &client_version);
+    if (res != 0)
+        return res;
+
+    // Store client_version in tags (in case supported_versions extension is absent)
+    set_tls_offered_version(tags, client_version);
+
+    if (client_version != TLS_VERSION12) {
```

Comment:
the version in the TLS header will never be 1.3. It goes up to 1.2 for backwards compatibility and then we try to find the "real" version in the extensions below, which can potentially contain 1.3.

Reply:
I will add a comment better describing this. We shouldn't change the condition because the logic below should only be evaluated if the header version is exactly 1.2. If it's 1.1 or less, there isn't any value in searching the extensions field in the payload

Reply:
added a comment on this logic